packages/lib/smart-account/src/lukso/LuksoKeyStore.ts

```diff
@@ -0,0 +1,97 @@
+import { SignedUserProfile } from '@dm3-org/dm3-lib-profile';
+import { stringify } from '@dm3-org/dm3-lib-shared';
+import { ethers } from 'ethers';
+import {
+    Dm3KeyStore,
+    IKeyStoreService as IKeyStore,
+} from '../KeyStore/IKeyStore';
+import { ERC725, ERC725JSONSchema } from '@erc725/erc725.js';
+import { ERC725JsonCoder } from './ERC725JsonCoder';
+
+const { toUtf8Bytes } = ethers.utils;
+
+declare global {
+    interface Window {
+        lukso?: any;
+    }
+}
+
+export class LuksoKeyStore implements IKeyStore {
+    //A wrapper around the UP contract
+    private readonly upContract: ethers.Contract;
+    private readonly erc725JsonCoder: ERC725JsonCoder;
+
+    constructor(upContract: ethers.Contract) {
+        this.upContract = upContract;
+
+        const erc725 = new ERC725(
+            ERC725JsonCoder.schemas,
+            this.upContract.address,
+            window.lukso!,
+        );
+        this.erc725JsonCoder = new ERC725JsonCoder(erc725);
+    }
+    getAccountAddress(): string {
+        return this.upContract.address;
+    }
+
+    async writeDm3Profile(userProfile: SignedUserProfile): Promise<void> {
+        const encoded = this.erc725JsonCoder.encodeDm3Profile(
+            this.upContract.address,
+            userProfile,
+        );
+
+        const { keys, values } = encoded;
+        await this.upContract.setDataBatch([...keys], [...values]);
+    }
+    async writeDm3KeyStore(keyStore: Dm3KeyStore): Promise<void> {
+        //Create a key for each address in the keyStore
+        const encodeDataReturn = Object.keys(keyStore).map((a) => {
+            return this.erc725JsonCoder.encodeDm3KeyStoreEntry(
+                a,
+                keyStore[a].signerPublicKey,
+                keyStore[a].encryptedProfileKeys,
+            );
+        });
+        //TODO find diff and post only modified keys
+        const keys = encodeDataReturn.map((a) => a.keys[0]);
+        const values = encodeDataReturn.map((a) => a.values[0]);
+
+        await this.upContract.setDataBatch(keys, values);
```

Comment:
await seems to be redundant here.

Reply:
I like that it is more descriptive than just knowing that the FN returns a Promise anyways. But we can change it ofc if its a concern for you